### PR TITLE
refactore operator `DifferenceTo*`

### PR DESCRIPTION
**Pull request by psychocoderHPC**
- refactore operator
- add documentation
- remove fixed 2D implementation by a generic one

**This pull request does not change that the difference is divided by the cell size (issue #820)**

~~**Please do not merge before #806**~~

### CODE REVIEW

**Comment by ax3l** [comment]
are there any implications on registers (we usually had issues on that part of the code, depending on the order of the differences).

File: src/picongpu/include/algorithms/DifferenceToLower.hpp [hunk 1]
```diff
@@ -28,59 +28,59 @@ namespace picongpu
...
-        case 1:
-            return (mem[0][0][0] - mem[0][-1][0]) * reciHeight;
-        case 2:
-            return (mem[0][0][0] - mem[-1][0][0]) * reciDepth;
-        }
-        return float3_X(NAN, NAN, NAN);
-    }
-};
+            const uint32_t direction = T_direction;
+            DataSpace<dim> dimension;
```
Comment (ax3l) [inline src/picongpu/include/algorithms/DifferenceToLower.hpp]: I would rename them to allow readers to identify the stencil again:

```c++
DataSpace<dim> indexIdentity;
DataSpace<dim> indexLower;
lowerIndex[direction] = -1;

return (mem(indexIdentity) - mem(indexLower)) / cellSize[direction];
```

**Comment by psychocoderHPC** [comment]
Yes it saves up to two register in some kernel. But I can't see any speedup which can be measured.

File: src/picongpu/include/algorithms/DifferenceToLower.hpp [hunk 2]
```diff
@@ -28,59 +28,59 @@ namespace picongpu
...
-template<>
-struct DifferenceToLower<DIM2>
-{
-    typedef PMacc::math::CT::Int<0, 0> OffsetEnd;
-    typedef PMacc::math::CT::Int<1, 1> OffsetOrigin;
 
-    template<class Memory >
-    HDINLINE typename Memory::ValueType operator()(const Memory& mem, const uint32_t direction) const
+    template<uint32_t T_direction>
+    struct GetDifference<T_direction, false>
```
Comment (ax3l) [inline src/picongpu/include/algorithms/DifferenceToLower.hpp]: pls let us add a coment why "dimension does not exist" == "constant in this direction" is a "d/dx = 0"

File: src/picongpu/include/algorithms/DifferenceToUpper.hpp [hunk 3]
```diff
@@ -27,59 +27,56 @@
...
-            return (mem[0][0][1] - mem[0][0][0]) * reciWidth;
-        case 1:
-            return (mem[0][1][0] - mem[0][0][0]) * reciHeight;
-        case 2:
-            return (mem[1][0][0] - mem[0][0][0]) * reciDepth;
-        }
-        return float3_X(NAN, NAN, NAN);
-    }
-};
+            const uint32_t direction = T_direction;
```
Comment (ax3l) [inline src/picongpu/include/algorithms/DifferenceToUpper.hpp]: DataSpace<dim> indexIdentity; /* defaults to (0, 0, 0) in 3D */
DataSpace<dim> indexUpper;    /* e.g., (0, 1, 0) for d/dy in 3D */
lowerIndex[direction] = 1;

return (mem(indexUpper) - mem(indexIdentity)) / cellSize[direction];

File: src/picongpu/include/fields/MaxwellSolver/Yee/Curl.hpp [hunk 4]
```diff
@@ -40,14 +38,10 @@ namespace yeeSolver
         template<class Memory >
         HDINLINE typename Memory::ValueType operator()(const Memory & mem) const
         {
-            Difference diff;
-            return float3_X(diff(mem, 1).z() - diff(mem, 2).y(),
-                               diff(mem, 2).x() - diff(mem, 0).z(),
-                               diff(mem, 0).y() - diff(mem, 1).x());
+            return float3_X(Difference::template GetDifference<1>()(mem).z() - Difference::template GetDifference<2>()(mem).y(),
```
Comment (ax3l) [inline src/picongpu/include/fields/MaxwellSolver/Yee/Curl.hpp]: let us typedef a bit to create a sexy formula with linear operators `D_{x,y,z}` from it:

```c++
typedef typename Difference::GetDifference<0> Dx;
typedef typename Difference::GetDifference<1> Dy;
typedef typename Difference::GetDifference<2> Dz;

return float3_X( Dy()(mem).z() - Dz()(mem).y(),
                 Dz()(mem).x() - Dx()(mem).z(),
                 Dx()(mem).y() - Dy()(mem).x());
```

File: src/picongpu/include/algorithms/DifferenceToLower.hpp [hunk 5]
```diff
@@ -28,59 +28,59 @@ namespace picongpu
...
+        HDINLINE typename Memory::ValueType operator()(const Memory& mem) const
         {
-        case 0:
-            return (mem[0][0] - mem[0][-1]) * reciWidth;
-        case 1:
-            return (mem[0][0] - mem[-1][0]) * reciHeight;
-        case 2:
-            return float3_X(0., 0., 0.);
-
+            return Memory::create(0.0);
```
Comment (ax3l) [inline src/picongpu/include/algorithms/DifferenceToLower.hpp]: probably ```Memory::ValueType::create(0.0)```

**Comment by ax3l** [comment]
Pretty cool anyway. :)

> Yes it saves up to two register in some kernel. But I can't see any
> speedup which can be measured.

**Comment by psychocoderHPC** [comment]
rebased

**Comment by ax3l** [comment]
great pull, thank you!

**Pull request merged**